Libraries/acdriver/Sources/CompileAction.cpp

```diff
@@ -27,10 +62,312 @@ CompileAction::
 {
 }
 
+static bool
+CompileAsset(
+    std::shared_ptr<xcassets::Asset::Catalog> const &catalog,
+    std::string ns,
+    Options const &options,
+    CompileOutput *compileOutput,
+    Result *result,
+    std::shared_ptr<xcassets::Asset::Asset> const &parent,
+    std::shared_ptr<xcassets::Asset::Asset> const &asset);
+
+template<typename T>
+static bool
+CompileChildren(
+    std::shared_ptr<xcassets::Asset::Catalog> const &catalog,
+    std::string ns,
+    Options const &options,
+    CompileOutput *compileOutput,
+    Result *result,
+    std::shared_ptr<xcassets::Asset::Asset> const &parent,
+    std::vector<T> const &assets)
+{
+    bool success = true;
+
+    for (auto const &asset : assets) {
+        if (!CompileAsset(catalog, ns, options, compileOutput, result, parent, asset)) {
+            success = false;
+        }
+    }
+
+    return success;
+}
+
+static std::string
+AssetReference(std::shared_ptr<xcassets::Asset::Asset> const &asset)
+{
+    // TODO: include [] for each key
+    return asset->path();
+}
+
+static void
+CompileAppIconSet(std::shared_ptr<xcassets::Asset::AppIconSet> const &appIconSet, CompileOutput *compileOutput, Result *result)
```

Comment:
This is incomplete enough it should probably be a separate PR after this one goes in, I think.